theano/misc/pkl_utils.py

```diff
@@ -93,3 +113,217 @@ class CompatUnpickler(pickle._Unpickler):
 else:
     class CompatUnpickler(pickle.Unpickler):
         pass
+
+
+class PersistentNdarrayID(object):
+    """Persist ndarrays in an object by saving them to a zip file.
+
+    :param zip_file: A zip file handle that the NumPy arrays will be saved to.
+    :type zip_file: :class:`zipfile.ZipFile`
+
+
+    .. note:
+        The convention for persistent ids given by this class and its derived
+        classes is that the name should take the form `type.name` where `type`
+        can be used by the persistent loader to determine how to load the
+        object, while `name` is human-readable and as descriptive as possible.
+
+    """
+    def __init__(self, zip_file):
+        self.zip_file = zip_file
+        self.count = 0
+        self.seen = {}
+
+    def _resolve_name(self, obj):
+        """Determine the name the object should be saved under."""
+        name = 'array_{}'.format(self.count)
+        self.count += 1
+        return name
+
+    def __call__(self, obj):
+        if ((type(obj) is numpy.ndarray) or
+                (type(obj) is cuda_ndarray.CudaNdarray)):
+            if id(obj) not in self.seen:
+                def write_array(f):
+                    numpy.lib.format.write_array(f, numpy.asarray(obj))
+                name = self._resolve_name(obj)
+                zipadd(write_array, self.zip_file, name)
+                if type(obj) is cuda_ndarray.CudaNdarray:
+                    self.seen[id(obj)] = 'cuda_ndarray.{}'.format(name)
+                else:
+                    self.seen[id(obj)] = 'ndarray.{}'.format(name)
+            return self.seen[id(obj)]
+
+
+class PersistentSharedVariableID(PersistentNdarrayID):
+    """Persist the names of shared variable arrays in the zip file.
```

Comment:
Right, this should probably be something like "Uses shared variable names when persisting to zip file."